CMakeLists.txt

```diff
@@ -80,7 +80,11 @@ IF(NOT KOKKOS_HAS_TRILINOS)
     # WORKAROUND FOR HIPCC
     IF(Kokkos_ENABLE_HIP)
       SET(KOKKOS_INTERNAL_CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
-      SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} --amdgpu-target=gfx906")
+      IF("${KOKKOS_CXX_COMPILER_ID}" STREQUAL Clang)
+        SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} --cuda-host-only")
+      ELSE()
+        SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} --amdgpu-target=gfx906")
```

Comment:
why are we hardcoding a GPU architecture here? Why isn't this set via the appropriate architecture flag?

Reply:
Not 100% sure on this (inherited it from of https://github.com/kokkos/kokkos/blob/develop/CMakeLists.txt#L83).  I believe the issue is roughly that with HCC, the `hipcc` script needs a `--amdgpu-target` or access to a GPU in order to even run `--version` / check validity. However, we don't pick an arch until `KOKKOS_SETUP_BUILD_ENVIRONMENT` on https://github.com/kokkos/kokkos/blob/develop/CMakeLists.txt#L157.  The new compiler actually respects the host-only flag, and doesn't require an arch flag to be specified, e.g.,:

```bash
cd /home/nick/kokkos/build/core/unit_test && /opt/rocm/bin/hipcc  -DGTEST_HAS_PTHREAD=0 -I/home/nick/kokkos/tpls/gtest -I/home/nick/kokkos/build/core/unit_test -I/home/nick/kokkos/core/unit_test  -std=gnu++11 --cuda-host-only -o CMakeFiles/kokkos_gtest.dir/__/__/tpls/gtest/gtest/gtest-all.cc.o -c /home/nick/coe0001/kokkos/tpls/gtest/gtest/gtest-all.cc
```


@dalg24 / @masterleinad / @Rombur -- can you comment further?

Reply:
Yes, this was introduced in https://github.com/kokkos/kokkos/pull/2792.
We need to give it some architecture to be able to get the version. Otherwise, we get
```
/opt/rocm-3.3.0/bin/hipcc     CMakeFiles/cmTC_10de7.dir/testCXXCompiler.cxx.o  -o cmTC_10de7 
    clang-10: error: ROCm agent detector could not identify any valid targets; please specify the target explicitly by passing a valid value to -amdgpu-target
```
or similar. At this point, we don't know the correct architecture when cross-compiling (unless we set some environment variables) so we set something arbitrary.

Reply:
Gotcha, for that we can can use the `--cuda-host-only` for the new compiler and avoid hard-coding a GPU target..  I think longer term it would be worth figuring out how to call the `clang++` compiler in `/opt/rocm/...` directly, to avoid the issues w/ the `hipcc` script (e.g., the one highlighted here) and make it irrelevant.